softioc/device.py

```diff
@@ -351,11 +351,29 @@ class ao(ProcessDeviceSupportOut):
     _ctype_ = c_double
     _dbf_type_ = fields.DBF_DOUBLE
 
+def to_epics_str_array(value):
+    """Convert the given array of Python strings to an array of EPICS
+    nul-terminated strings"""
+    result = numpy.empty(len(value), 'S40')
+
+    for n, s in enumerate(value):
+        if isinstance(s, str):
+            val = EpicsString._ctype_()
+            val.value = s.encode() + b'\0'
+            result[n] = val.value
+        else:
+            result[n] = s
+    return result
+
 
 def _require_waveform(value, dtype):
     if isinstance(value, bytes):
         # Special case hack for byte arrays.  Surprisingly tricky:
         value = numpy.frombuffer(value, dtype = numpy.uint8)
+
+    if dtype and dtype.char == 'S':
+        return to_epics_str_array(value)
+
```

Comment:
Not a huge fan of early return, and looking at the context this is a case for using `else:`.

Also, this test should probably be done _before_ the `bytes` check above, making this a stronger case for being an `if ... else ...` block.

Reply:
Rewritten in 5b59ed7f86a126842467247dfb004ecf21aed561